src/Microsoft.ComponentDetection.Detectors/rust/RustCliDetector.cs

```diff
@@ -0,0 +1,134 @@
+namespace Microsoft.ComponentDetection.Detectors.Rust;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.ComponentDetection.Contracts;
+using Microsoft.ComponentDetection.Contracts.Internal;
+using Microsoft.ComponentDetection.Contracts.TypedComponent;
+using Microsoft.ComponentDetection.Detectors.Rust.Contracts;
+using Microsoft.Extensions.Logging;
+
+/// <summary>
+/// A Rust CLI detector that uses the cargo metadata command to detect Rust components.
+/// </summary>
+public class RustCliDetector : FileComponentDetector, IExperimentalDetector
+{
+    private readonly ICommandLineInvocationService cliService;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RustCliDetector"/> class.
+    /// </summary>
+    /// <param name="componentStreamEnumerableFactory">The component stream enumerable factory.</param>
+    /// <param name="walkerFactory">The walker factory.</param>
+    /// <param name="cliService">The command line invocation service.</param>
+    /// <param name="logger">The logger.</param>
+    public RustCliDetector(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        ICommandLineInvocationService cliService,
+        ILogger<RustCliDetector> logger)
+    {
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.Scanner = walkerFactory;
+        this.cliService = cliService;
+        this.Logger = logger;
+    }
+
+    /// <inheritdoc />
+    public override string Id => "RustCli";
+
+    /// <inheritdoc />
+    public override IEnumerable<string> Categories { get; } = new[] { "Rust" };
+
+    /// <inheritdoc />
+    public override IEnumerable<ComponentType> SupportedComponentTypes => new[] { ComponentType.Cargo };
+
+    /// <inheritdoc />
+    public override int Version => 1;
+
+    /// <inheritdoc />
+    public override IList<string> SearchPatterns { get; } = new[] { "Cargo.toml" };
+
+    /// <inheritdoc />
+    protected override async Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+    {
+        var componentStream = processRequest.ComponentStream;
+
+        try
+        {
+            if (!await this.cliService.CanCommandBeLocatedAsync("cargo", null))
+            {
+                this.Logger.LogWarning("Could not locate cargo command. Skipping Rust CLI detection");
+                return;
+            }
+
+            var cliResult = await this.cliService.ExecuteCommandAsync(
```

Comment:
This can fail. Should be handled / could kick off the existing parse-based detector if it does.

Reply:
We dont have that functionality untill we merge the detectors.